File: js/src/library/runtimeScope.js
import { applyObject } from './apply';
import RuntimeSignal from './runtimeSignal';

/**
 * Runtime scope for managing contextual runtime information.
 *
 * Fields of the scope must not be modified directly.
 * Instead, `JPLRuntimeScope.next` should be used to create a sub scope with the specified modifications.
 */
class JPLRuntimeScope {
  constructor(presets = {}) {
    this._state = {
      signal: presets.signal ?? new RuntimeSignal(),
      vars: presets.vars ?? {},
    };
  }

  get signal() {
    return this._state.signal;
  }

  get vars() {
    return this._state.vars;
  }

  /** Inherit the next scope based on the specified modifications */
  next(modifications) {
    return new JPLRuntimeScope({
      signal: modifications.signal ?? this.signal,
      vars: applyObject(this.vars, Object.entries(modifications.vars ?? {})),
    });
  }
}

export default JPLRuntimeScope;

File: go/library/RuntimeScope.go
package library

import "github.com/jplorg/jpl/go/v2/jpl"

func NewRuntimeScope(presets *jpl.JPLRuntimeScopeConfig) jpl.JPLRuntimeScope {
	if presets == nil {
		presets = new(jpl.JPLRuntimeScopeConfig)
	}

	signal := presets.Signal
	if signal == nil {
		signal = NewRuntimeSignal(nil)
	}

	return &runtimeScope{
		signal: signal,
		vars:   presets.Vars,
	}
}

type runtimeScope struct {
	signal jpl.JPLRuntimeSignal
	vars   map[string]any
}

func (s *runtimeScope) Signal() jpl.JPLRuntimeSignal {
	return s.signal
}

func (s *runtimeScope) Vars() map[string]any {
	return s.vars
}

func (s *runtimeScope) Next(modifications *jpl.JPLRuntimeScopeConfig) jpl.JPLRuntimeScope {
	if modifications == nil {
		modifications = new(jpl.JPLRuntimeScopeConfig)
	}

	signal := modifications.Signal
	if signal == nil {
		signal = s.signal
	}

	return &runtimeScope{
		signal: signal,
		vars:   ApplyObject(s.vars, ObjectEntries(modifications.Vars)),
	}
}
